### PR TITLE
Added docdbClusterBackupRetention plugin and spec

### PR DESCRIPTION
**Pull request by m-nouman-suleman**
## Task
- https://scalock.atlassian.net/browse/SAAS-3181

### CODE REVIEW

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    recommended_action: 'Modify DocumentDb cluster to configure sufficient backup retention period.',
+    link: 'https://docs.aws.amazon.com/documentdb/latest/developerguide/db-cluster-modify.html',
+    apis: ['DocDB:describeDBClusters'],
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+
+        var RECOMMENDED_THRESHOLD = 7;
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:21]: We should add a setting for this. Default can be 7. Regex should only allow 1-35

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'DocumentDB Cluster Backup Retention',
+    category: 'DocumentDB',
+    domain: 'Databases',
+    description: 'Ensure that your Amazon DocumentDB clusters have set a minimum backup retention period.',
+    more_info: `DocumentDB cluster provides feature to retain incremental backups between 1 and 35 allowing
+                you to quickly restore to any point within the backup retention period. Ensure that you have
+                sufficient backup retention period configured in order to restore your data in the event of failure.`,
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:11]:
```suggestion
    more_info: 'DocumentDB cluster provides feature to retain incremental backups between 1 and 35 allowing you to quickly restore to any point within the backup retention period. Ensure that you have sufficient backup retention period configured in order to restore your data in the event of failure.',
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            regex: '^[1-35]*$',
+            default: 7
+        }
+    },
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+        var doc_db_backup_retention_threshold = settings.doc_db_backup_retention_threshold || this.settings.doc_db_backup_retention_threshold.default;
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:26]:
```suggestion
        var doc_db_backup_retention_threshold = parseInt(settings.doc_db_backup_retention_threshold || this.settings.doc_db_backup_retention_threshold.default); 
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    category: 'DocumentDB',
+    domain: 'Databases',
+    description: 'Ensure that your Amazon DocumentDB clusters have set a minimum backup retention period.',
+    more_info: 'DocumentDB cluster provides feature to retain incremental backups between 1 and 35 allowing you to quickly restore to any point within the backup retention period. Ensure that you have sufficient backup retention period configured in order to restore your data in the event of failure.',
+    recommended_action: 'Modify DocumentDb cluster to configure sufficient backup retention period.',
+    link: 'https://docs.aws.amazon.com/documentdb/latest/developerguide/db-cluster-modify.html',
+    apis: ['DocDB:describeDBClusters'],
+    settings: {
+        doc_db_backup_retention_threshold: {
+            name: 'Threshold for DocDB Cluster backup retention.',
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:15]:
```suggestion
            name: 'DocDB Cluster Backup Retention Threshold',
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            
+            for (let cluster of describeDBClusters.data) {
+                if (!cluster.DBClusterArn) continue;
+
+                let resource = cluster.DBClusterArn;
+
+                if (cluster.BackupRetentionPeriod && cluster.BackupRetentionPeriod > doc_db_backup_retention_threshold) {
+                    helpers.addResult(results, 0,
+                        `DocumentDB cluster has a backup retention period of ${cluster.BackupRetentionPeriod} days \
+                        which is greater than or equal to the recommended period of ${doc_db_backup_retention_threshold} days`,
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:54]:
```suggestion
                        `DocumentDB cluster has a backup retention period of ${cluster.BackupRetentionPeriod} of ${doc_db_backup_retention_threshold} days limit`,
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                        region, resource);
+                } else {
+                    helpers.addResult(results, 2,
+                        `DocumentDB cluster has a backup retention period of ${cluster.BackupRetentionPeriod} days \
+                        which is less than the recommended period of ${doc_db_backup_retention_threshold} days`,
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:58]:
```suggestion
                        `DocumentDB cluster has a backup retention period of ${cluster.BackupRetentionPeriod} of ${doc_db_backup_retention_threshold} days limit`,
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/documentDB/docdbClusterBackupRetention.spec.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                done();
+            });
+        });
+
+        it('should FAIL if DocumentDB Clusters do not have the recommended backup retention period', function (done) {
+            const cache = createCache([describeDBClusters[0]]);
+            docdbClusterBackupRetention.run(cache, { doc_db_backup_retention_threshold: 7 }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('DocumentDB cluster has a backup retention period of 1 days');
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.spec.js:100]:
```suggestion
                expect(results[0].message).to.include('DocumentDB cluster has a backup retention period of 1');
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.spec.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+};
+
+describe('docdbClusterBackupRetention', function () {
+    describe('run', function () {
+        it('should PASS if DocumentDb Cluster has the recommended backup retention period', function (done) {
+            const cache = createCache([describeDBClusters[1]]);
+            docdbClusterBackupRetention.run(cache, { doc_db_backup_retention_threshold: 7 }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('DocumentDB cluster has a backup retention period of 10 days');
```
Comment (AkhtarAmir) [inline plugins/aws/documentDB/docdbClusterBackupRetention.spec.js:89]:
```suggestion
                expect(results[0].message).to.include('DocumentDB cluster has a backup retention period of 10');
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    domain: 'Databases',
+    description: 'Ensure that your Amazon DocumentDB clusters have set a minimum backup retention period.',
+    more_info: 'DocumentDB cluster provides feature to retain incremental backups between 1 and 35 allowing you to quickly restore to any point within the backup retention period. Ensure that you have sufficient backup retention period configured in order to restore your data in the event of failure.',
+    recommended_action: 'Modify DocumentDb cluster to configure sufficient backup retention period.',
+    link: 'https://docs.aws.amazon.com/documentdb/latest/developerguide/db-cluster-modify.html',
+    apis: ['DocDB:describeDBClusters'],
+    settings: {
+        doc_db_backup_retention_threshold: {
+            name: 'DocDB Cluster Backup Retention Threshold',
+            description: 'Plugin results will become aggregated once this value is breached',
```
Comment (dipsubha06) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:16]:
```suggestion
            description: 'Desired number of days for DocumentDB cluster backup retention period.',
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                    'No DocumentDB clusters found', region);
+                return rcb();
+            }
+            
+            for (let cluster of describeDBClusters.data) {
+                if (!cluster.DBClusterArn) continue;
+
+                let resource = cluster.DBClusterArn;
+
+                if (cluster.BackupRetentionPeriod && cluster.BackupRetentionPeriod > doc_db_backup_retention_threshold) {
```
Comment (dipsubha06) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:51]:
```suggestion
                if (cluster.BackupRetentionPeriod && cluster.BackupRetentionPeriod >=  doc_db_backup_retention_threshold) {
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    category: 'DocumentDB',
+    domain: 'Databases',
+    description: 'Ensure that your Amazon DocumentDB clusters have set a minimum backup retention period.',
+    more_info: 'DocumentDB cluster provides feature to retain incremental backups between 1 and 35 allowing you to quickly restore to any point within the backup retention period. Ensure that you have sufficient backup retention period configured in order to restore your data in the event of failure.',
+    recommended_action: 'Modify DocumentDb cluster to configure sufficient backup retention period.',
+    link: 'https://docs.aws.amazon.com/documentdb/latest/developerguide/db-cluster-modify.html',
+    apis: ['DocDB:describeDBClusters'],
+    settings: {
+        doc_db_backup_retention_threshold: {
+            name: 'DocDB Cluster Backup Retention Threshold',
```
Comment (dipsubha06) [inline plugins/aws/documentDB/docdbClusterBackupRetention.js:15]:
```suggestion
            name: 'DocDB Cluster Minimum Backup Retention Period',
```

File: plugins/aws/documentDB/docdbClusterBackupRetention.spec.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('No DocumentDB clusters found');
+                expect(results[0].region).to.equal('us-east-1');
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to list DocumentDB Clusters', function (done) {
+            const cache = createCache(null, null, null, { message: "Unable to list DocumentDB Clusters encryption" });
```
Comment (dipsubha06) [inline plugins/aws/documentDB/docdbClusterBackupRetention.spec.js:118]:
```suggestion
            const cache = createCache(null,  { message: "Unable to list DocumentDB Clusters encryption" });
```

**Review by dipsubha06 — APPROVED** [review]

**Pull request merged**